modules/javafx.graphics/src/main/java/com/sun/javafx/scene/paint/PaintUtils.java

```diff
@@ -0,0 +1,118 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.scene.paint;
+
+import javafx.scene.paint.Color;
+import javafx.scene.paint.ImagePattern;
+import javafx.scene.paint.LinearGradient;
+import javafx.scene.paint.Paint;
+import javafx.scene.paint.RadialGradient;
+import javafx.scene.paint.Stop;
+import java.util.List;
+
+public final class PaintUtils {
+
+    private PaintUtils() {}
+
+    /**
+     * Interpolates between potentially different types of paint.
+     * <p>
+     * In addition to homogeneous interpolations between paints of the same type, the following
+     * heterogeneous interpolations are supported:
+     * <ul>
+     *     <li>Color ↔ LinearGradient
+     *     <li>Color ↔ RadialGradient
+     * </ul>
+     * If a paint is not interpolatable, {@code startValue} is returned for {@code t < 0.5},
+     * and {@code endValue} is returned otherwise.
+     */
+    public static Paint interpolate(Paint startValue, Paint endValue, double t) {
+        if (startValue instanceof Color start) {
+            if (endValue instanceof Color end) {
+                return start.interpolate(end, t);
+            }
+
+            if (endValue instanceof LinearGradient end) {
+                return newSolidGradient(end, start).interpolate(end, t);
+            }
+
+            if (endValue instanceof RadialGradient end) {
+                return newSolidGradient(end, start).interpolate(end, t);
+            }
+        }
+
+        if (startValue instanceof LinearGradient start) {
+            if (endValue instanceof LinearGradient end) {
+                return start.interpolate(end, t);
+            }
+
+            if (endValue instanceof Color end) {
+                return start.interpolate(newSolidGradient(start, end), t);
+            }
+        }
+
+        if (startValue instanceof RadialGradient start) {
+            if (endValue instanceof RadialGradient end) {
+                return start.interpolate(end, t);
+            }
+
+            if (endValue instanceof Color end) {
+                return start.interpolate(newSolidGradient(start, end), t);
+            }
+        }
+
+        if (startValue instanceof ImagePattern start && endValue instanceof ImagePattern end) {
+            return start.interpolate(end, t);
+        }
+
+        return t < 0.5 ? startValue : endValue;
+    }
+
+    /**
+     * Creates a new linear gradient that consists of two stops with the same color.
+     */
+    public static LinearGradient newSolidGradient(LinearGradient source, Color color) {
+        return new LinearGradient(
+                source.getStartX(), source.getStartY(),
+                source.getEndX(), source.getEndY(),
+                source.isProportional(),
+                source.getCycleMethod(),
```

Comment:
We can't interpolate between relative and absolute gradients (the interpolate method lacks the context necessary to convert between the two), so we need to copy these from source instead of always choosing one or the other.